sgl-kernel/csrc/cpu/mamba/fla.cpp

```diff
@@ -0,0 +1,1068 @@
+        auto curr_k_pad = k_pad_ptr + j * qk_head_size;
+        int64_t i = 0;
+        auto vec_zero = at::vec::Vectorized<scalar_t>(0.0);
+        for (; i < reduced_vec_size * (qk_head_size / reduced_vec_size); i += reduced_vec_size) {
+          vec_zero.store(curr_q_pad + i);
+          vec_zero.store(curr_k_pad + i);
+        }
+        if (i < qk_head_size) {
+          vec_zero.store(curr_q_pad + i, qk_head_size - i);
+          vec_zero.store(curr_k_pad + i, qk_head_size - i);
+        }
+        for (auto hi = 0; hi < head_group; hi++) {
+          int64_t h = h_qk * head_group + hi;
+          auto curr_v_pad = v_pad_ptr + h * global_total_seq_length * v_head_size + j * v_head_size;
+          auto curr_k_beta = k_beta_ptr + h * global_total_seq_length * qk_head_size + j * qk_head_size;
+          auto curr_v_beta = v_beta_ptr + h * global_total_seq_length * v_head_size + j * v_head_size;
+          int64_t i = 0;
+          for (; i < reduced_vec_size * (qk_head_size / reduced_vec_size); i += reduced_vec_size) {
+            vec_zero.store(curr_k_beta + i);
+          }
+          if (i < qk_head_size) {
+            vec_zero.store(curr_k_beta + i, qk_head_size - i);
+          }
+          i = 0;
+          for (; i < reduced_vec_size * (v_head_size / reduced_vec_size); i += reduced_vec_size) {
+            vec_zero.store(curr_v_pad + i);
+            vec_zero.store(curr_v_beta + i);
+          }
+          if (i < v_head_size) {
+            vec_zero.store(curr_v_pad + i, v_head_size - i);
+            vec_zero.store(curr_v_beta + i, v_head_size - i);
+          }
+        }
+      }
+      // Move to the next query
+      data_index_step(h_qk, qk_num_head, c, global_num_chunk);
+    }
+  });
+
+  at::parallel_for(0, v_num_head * global_num_chunk, 1, [&](int64_t begin, int64_t end) {
+    int64_t h = 0, c = 0;
+    data_index_init(begin, h, v_num_head, c, global_num_chunk);
+    int ompIdx = at::get_thread_num();
+    int64_t offset = 0;
+    scalar_t* thread_buff_ptr = thread_buff + ompIdx * buff_size_16bit_per_thread;
+    scalar_t* k_transpose = thread_buff_ptr + offset;
+    offset += qk_head_size * chunk_size;
+    scalar_t* v_pack = thread_buff_ptr + offset;
+    offset += chunk_size * v_head_size;
+    scalar_t* k_beta_g = thread_buff_ptr + offset;
+    offset += chunk_size * qk_head_size;
+    scalar_t* k_beta_g_pack = thread_buff_ptr + offset;
+    offset += chunk_size * qk_head_size;
+    float* curr_attn = reinterpret_cast<float*>(thread_buff_ptr + offset);
+    offset += chunk_size * chunk_size * 2;
+    scalar_t* curr_attn_reduced = thread_buff_ptr + offset;
+    offset += chunk_size * chunk_size;
+    float* k_cumdecay = reinterpret_cast<float*>(thread_buff_ptr + offset);
+    offset += chunk_size * qk_head_size * 2;
+    float* row = reinterpret_cast<float*>(thread_buff_ptr + offset);
+    offset += chunk_size * 2;
+    float* updated = reinterpret_cast<float*>(thread_buff_ptr + offset);
```

Comment:
```suggestion
THREAD_BUFFER_ALLOC(k_transpose, thread_buff_ptr, offset, scalar_t, qk_head_size * chunk_size)
THREAD_BUFFER_ALLOC(v_pack, thread_buff_ptr, offset, scalar_t, chunk_size * v_head_size)
THREAD_BUFFER_ALLOC(k_beta_g, thread_buff_ptr, offset, scalar_t, chunk_size * qk_head_size)
THREAD_BUFFER_ALLOC(k_beta_g_pack, thread_buff_ptr, offset, scalar_t, chunk_size * qk_head_size)
THREAD_BUFFER_ALLOC(curr_attn, thread_buff_ptr, offset, float, chunk_size * chunk_size * 2)
THREAD_BUFFER_ALLOC(curr_attn_reduced, thread_buff_ptr, offset, scalar_t, chunk_size * chunk_size)
THREAD_BUFFER_ALLOC(k_cumdecay, thread_buff_ptr, offset, float, chunk_size * qk_head_size * 2)
THREAD_BUFFER_ALLOC(row, thread_buff_ptr, offset, float, chunk_size * 2)
THREAD_BUFFER_ALLOC(updated, thread_buff_ptr, offset, float, chunk_size * 2)
```

OK, this is kind of like your personal style :) we can use a MACRO to simplify it a little bit.